### PR TITLE
[fix)(colocate join) fix wrong use of colocate join

### LINKED ISSUE #36324: [Bug] wrong colocate join used
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Version

nereids planner has this problem, all versions.

### What's Wrong?

colocate group keys do not the same as on condition, we should no use colocate join, then make the result do not correctly.

### What You Expected?

get correct result

### How to Reproduce?

1. create table a, use k1 as distributed key
```
CREATE TABLE `a` (
  `k1` INT NOT NULL,
  `k2` INT NOT NULL,
  `v` INT SUM NULL
) ENGINE=OLAP
AGGREGATE KEY(`k1`, `k2`)
COMMENT 'OLAP'
DISTRIBUTED BY HASH(`k1`) BUCKETS 10
PROPERTIES (
"replication_allocation" = "tag.location.default: 1",
"colocate_with" = "k1_group"
); 
```

2. create table b, use k2 as distributed key
```
CREATE TABLE `b` (
  `k1` INT NOT NULL,
  `k2` INT NOT NULL,
  `v` INT SUM NULL
) ENGINE=OLAP
AGGREGATE KEY(`k1`, `k2`)
COMMENT 'OLAP'
DISTRIBUTED BY HASH(`k2`) BUCKETS 10
PROPERTIES (
"replication_allocation" = "tag.location.default: 1",
"colocate_with" = "k1_group"
); 
```

3. and table a and b use the same colocate group: `k1_group`
4. add some data to table and table b
```
insert into a values(1,2,3);
insert into b values(1,2,3);
```
5. check the plan, we found colocate join is used. Actually in this case, we should not use colocate join:
```
> explain select * from a join b on a.k1=b.k1 and a.k2=b.k2;
+----------------------------------------------------------------------------------+
| Explain String(Nereids Planner)                                                  |
+----------------------------------------------------------------------------------+
| PLAN FRAGMENT 0                                                                  |
|   OUTPUT EXPRS:                                                                  |
|     k1[#12]                                                                      |
|     k2[#13]                                                                      |
|     v[#14]                                                                       |
|     k1[#15]                                                                      |
|     k2[#16]                                                                      |
|     v[#17]                                                                       |
|   PARTITION: UNPARTITIONED                                                       |
|                                                                                  |
|   HAS_COLO_PLAN_NODE: false                                                      |
|                                                                                  |
|   VRESULT SINK                                                                   |
|      MYSQL_PROTOCAL                                                              |
|                                                                                  |
|   3:VEXCHANGE                                                                    |
|      offset: 0                                                                   |
|      distribute expr lists: k1[#12]                                              |
|                                                                                  |
| PLAN FRAGMENT 1                                                                  |
|                                                                                  |
|   PARTITION: HASH_PARTITIONED: k1[#3]                                            |
|                                                                                  |
|   HAS_COLO_PLAN_NODE: true                                                       |
|                                                                                  |
|   STREAM DATA SINK                                                               |
|     EXCHANGE ID: 03                                                              |
|     UNPARTITIONED                                                                |
|                                                                                  |
|   2:VHASH JOIN(136)                                                              |
|   |  join op: INNER JOIN(COLOCATE[])[]                                           |
|   |  equal join conjunct: (k1[#3] = k1[#0])                                      |
|   |  equal join conjunct: (k2[#4] = k2[#1])                                      |
|   |  cardinality=1                                                               |
|   |  vec output tuple id: 3                                                      |
|   |  output tuple id: 3                                                          |
|   |  vIntermediate tuple ids: 2                                                  |
|   |  hash output slot ids: 0 1 2 3 4 5                                           |
|   |  final projections: k1[#6], k2[#7], v[#8], k1[#9], k2[#10], v[#11]           |
|   |  final project output tuple id: 3                                            |
|   |  distribute expr lists: k1[#3]                                               |
|   |  distribute expr lists: k2[#1]                                               |
|   |                                                                              |
|   |----0:VOlapScanNode(133)                                                      |
|   |       TABLE: doris0.b(b), PREAGGREGATION: OFF. Reason: No aggregate on scan. |
|   |       partitions=1/1 (b)                                                     |
|   |       tablets=10/10, tabletList=30133,30135,30137 ...                        |
|   |       cardinality=0, avgRowSize=0.0, numNodes=1                              |
|   |       pushAggOp=NONE                                                         |
|   |                                                                              |
|   1:VOlapScanNode(132)                                                           |
|      TABLE: doris0.a(a), PREAGGREGATION: OFF. Reason: No aggregate on scan.      |
|      partitions=1/1 (a)                                                          |
|      tablets=10/10, tabletList=30082,30084,30086 ...                             |
|      cardinality=0, avgRowSize=0.0, numNodes=1                                   |
|      pushAggOp=NONE                                                              |
|                                                                                  |
|                                                                                  |
| Statistics                                                                       |
|  planed with unknown column statistics                                           |
+----------------------------------------------------------------------------------+
```

### Anything Else?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by cambyzju**
## Proposed changes

Issue Number: close https://github.com/apache/doris/issues/36324

### CODE REVIEW

**Comment by cambyzju** [comment]
run buildall

**Comment by cambyzju** [comment]
run buildall

**Comment by cambyzju** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Review by morrySnow — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java [hunk 1]
```diff
@@ -300,12 +301,42 @@ public static boolean couldColocateJoin(DistributionSpecHash leftHashSpec, Distr
...
+        }
+
+        Set<Integer> equalIndices = new HashSet<>();
+        for (Expression expr : conjuncts) {
+            // only simple equal predicate can use colocate join
+            if (!(expr instanceof EqualPredicate)) {
+                return false;
+            }
+            Expression leftChild = ((EqualPredicate) expr).left();
+            Expression rightChild = ((EqualPredicate) expr).right();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:319]: not always left for left child, right for right child

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java [hunk 2]
```diff
@@ -300,12 +301,42 @@ public static boolean couldColocateJoin(DistributionSpecHash leftHashSpec, Distr
...
+            if (!(leftChild instanceof Slot) || !(rightChild instanceof Slot)) {
+                return false;
+            }
+
+            // on conditions must keep same order as distributed columns
+            Integer leftIndex = leftHashSpec.getExprIdToEquivalenceSet().get(((Slot) leftChild).getExprId());
+            Integer rightIndex = rightHashSpec.getExprIdToEquivalenceSet().get(((Slot) rightChild).getExprId());
+            if (leftIndex != rightIndex) {
```
[the remark this answers is not in the document]
Comment (cambyzju) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:327]: Got, I will fix it later

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java (hunk 1 above)
Comment (cambyzju) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:319]: Got it, maybe it is the reason why lots of regression tests of tcpds + tpch shape failed, I will try to fix it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (cambyzju) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java (hunk 1 above)
Comment (cambyzju) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:321]: done

**Comment by cambyzju** [comment]
run buildall

**Review by 924060929 — APPROVED** [review]

**Review by Yulei-Yang — APPROVED** [review]

**Pull request merged**